airflow/www_rbac/decorators.py

```diff
@@ -91,3 +93,47 @@ def zipper(response):
         return f(*args, **kwargs)
 
     return view_func
+
+
+def has_dag_access(**dag_kwargs):
+    """
+    Decorator to check whether the user has read / write permission on the dag.
+    """
+    def decorator(f):
+        @functools.wraps(f)
+        def wrapper(self, *args, **kwargs):
+            dag_id = request.args.get('dag_id')
+            # if it is true, we need to check whether user has read access on the dag
+            can_dag_read = dag_kwargs.get('can_dag_read', False)
+            # if it is false, we need to check whether user has write access on the dag
+            can_dag_edit = dag_kwargs.get('can_dag_edit', False)
+            if dag_id is None or (not can_dag_edit and not can_dag_read):
+                # this request is not dag-level related
+                # or no read/write level access perm required
+                return f(self, *args, **kwargs)
+
+            # check whether the role of user has all_dag write perm
+            # (no need to check read perm)
```

Comment:
NIT: lots of calls to long `self.appbuilder.sm.has_access`, you could alias it with `has_access = self.appbuilder.sm.has_access` and use this shorter call many times

Reply:
Looks like this block of code can be simplified quite a bit. How about:
```python
has_access = self.appbuilder.sm.has_access
if (
    has_access('can_dag_edit', 'all_dags') or
    has_access('can_dag_edit', dag_id) or (
        not can_dag_edit and (has_access('can_dag_read', 'all_dags') or has_access('can_dag_read', dag_id)),
    )
):
    return f(self, *args, **kwargs)
else:
    return redirect(url_for(self.appbuilder.sm.auth_view.__class__.__name__ + ".login"))
```

Reply:
thanks. much better :)